src/Validator/Authentication.php

```diff
@@ -189,61 +200,71 @@ public function setService(AuthenticationService $service)
     }
 
     /**
-     * Is Valid
+     * Returns true if and only if authentication result is valid
+     *
+     * If authentication result fails validation, then this method returns false, and
+     * getMessages() will return an array of messages that explain why the
+     * validation failed.
      *
-     * @param  mixed $value
-     * @param  array $context
+     * @param  mixed $value   OPTIONAL Credential (or field)
+     * @param  array $context OPTIONAL Authentication data (identity and/or credential)
      * @return bool
+     * @throws Exception\RuntimeException
      */
     public function isValid($value = null, $context = null)
     {
         if ($value !== null) {
             $this->setCredential($value);
         }
 
+        if ($this->identity === null) {
+            throw new Exception\RuntimeException('Identity must be set prior to validation');
+        }
         if (($context !== null) && array_key_exists($this->identity, $context)) {
             $identity = $context[$this->identity];
         } else {
             $identity = $this->identity;
         }
-        if (!$this->identity) {
-            throw new Exception\RuntimeException('Identity must be set prior to validation');
-        }
 
+        if ($this->credential === null) {
+            throw new Exception\RuntimeException('Credential must be set prior to validation');
+        }
         if (($context !== null) && array_key_exists($this->credential, $context)) {
             $credential = $context[$this->credential];
         } else {
             $credential = $this->credential;
         }
 
-        if (!$this->adapter) {
-            throw new Exception\RuntimeException('Adapter must be set prior to validation');
-        }
-        $this->adapter->setIdentity($identity);
-        $this->adapter->setCredential($credential);
-
         if (!$this->service) {
             throw new Exception\RuntimeException('AuthenticationService must be set prior to validation');
         }
+
+        if (!$this->adapter) {
+            $adapter = $this->service->getAdapter();
+            if (!$adapter) {
+                throw new Exception\RuntimeException('Adapter must be set prior to validation');
+            }
+            if (!$adapter instanceof ValidatableAdapterInterface) {
+                throw new Exception\RuntimeException(sprintf(
+                    'Adapter must be an instance of ValidatableAdapterInterface, %s given',
+                    (is_object($adapter) ? get_class($adapter) : gettype($adapter))
+                ));
+            }
+        } else {
+            $adapter = $this->adapter;
+        }
```

Comment:
Extract the logic for pulling the adapter from the `AuthenticationService` to a new method, so you can then instead say `$adapter = $this->adapter ?: $this->getAdapterFromAuthenticationService();`.

I'll make that change in a separate commit that I push to your branch.